### ISSUE TITLE
Next.js sets invalid and irrelevant tsconfig options.

### ISSUE THREAD

**Issue by remcohaszing** [vercel/next.js#45617]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
    Operating System:
      Platform: linux
      Arch: x64
      Version: #202212290932~1674139725~22.04~ca93ccf SMP PREEMPT_DYNAMIC Thu J
    Binaries:
      Node: 19.6.0
      npm: 9.4.1
      Yarn: 1.22.19
      pnpm: 7.13.5
    Relevant packages:
      next: 13.1.7-canary.5
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

TypeScript

### Link to the code that reproduces this issue

https://stackblitz.com/edit/vercel-next-js-uxjbhb?file=tsconfig.json

### To Reproduce

- Open the StackBlitz link
- Open `tsconfig.json`

### Describe the Bug

Next.js generates undesirable `tsconfig.json` options. Some options are undesirable or unnecessary, others are wrong.

### Expected Behavior

Based on the reasoning emitted by Next.js:

> The following suggested values were added to your `tsconfig.json`. These values **can be changed** to fit your project's needs:
>
> - lib was set to **`dom,dom.iterable,esnext`**

This option matches the `target` if it is defined. I want to omit this.

> - allowJs was set to **`true`**

This option is true if `checkJs` is true. I want to omit this.

> - skipLibCheck was set to **`true`**

This definitely makes sense when generating a default `tsconfig.json`, but there’s not need to enforce this.

> - strict was set to **`false`**

This option makes sense to me.

> - forceConsistentCasingInFileNames was set to **`true`**

This option makes sense to me. Note that this will be default in TypeScript 5. From that version on I want to omit this.

> - noEmit was set to **`true`**

This option makes sense to me.

> - incremental was set to **`true`**

This option makes sense to me.

> - include was set to **`['next-env.d.ts', '**/*.ts', '**/*.tsx']`**

All of these files are included by default. Also `.js` / `.jsx` files are enabled by default if `checkJs` is true. I want to omit this.

> - exclude was set to **`['node_modules']`**

This is default behaviour. I want to omit this.

> The following **mandatory changes** were made to your `tsconfig.json`:
>
> - esModuleInterop was set to **`true`** (requirement for SWC / babel)

This option is only relevant for CJS. I use `"type": "module"` in `package.json`, so this option should be removed.

> - module was set to **`esnext`** (for dynamic import() support)

The correct value if `"type": "module"` in `package.json` is `node16`, not `esnext`.

> - moduleResolution was set to **`node`** (to match webpack resolution)

The correct value if `"type": "module"` in `package.json` is `node16`, not `node`. This is default behaviour when `module` is set to `node16`, so it should be removed.

> - resolveJsonModule was set to **`true`** (to match webpack resolution)

This definitely makes sense when generating a default `tsconfig.json`, but there’s not need to enforce this.

> - isolatedModules was set to **`true`** (requirement for SWC / Babel)

This option makes sense to me personally, but it’s a preference. Users should be able to omit this.

> - jsx was set to **`preserve`** (next.js implements its own optimized jsx transform)

The reasoning makes sense for compiling, but for typechecking purposes, this should match the runtime (classic or automatic) used.

**Comment by joshjg** [comment]
`moduleResolution` should be "bundler" to match webpack behavior.

**Comment by Kurt-von-Laven** [comment]
I agree with the overall theme of the issue although not all the details of which options are wrong since that is relative to any given project's needs and preferences. For instance, we don't want to set `noEmit: true` in our project. Overall, it would be great either to have a flag that prevents `next build` from modifying `tsconfig.json` and/or to stop modifying preferences that aren't strictly necessary for Next.js to work properly.

**Comment by Oren-Schmerling** [comment]
I'm working on a project where it sets `"noEmit": true, "incremental": true`. Neither of which I want. Is there still no way to disable this?

**Comment by Blitz2145** [comment]
This is still happening, and as of typescript v6, strict is a default but nextjs still tries rewriting the user's tsconfig.json for them to insert it. This is not helpful behavior, at most nextjs should issue a warning for an incorrect tsconfig. Just like with linting, this is outside the scope of what nextjs should be doing at all imo.

_Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

**Comment by jramstedt** [comment]
[TypeScript 6.0](https://devblogs.microsoft.com/typescript/announcing-typescript-6-0/) now has "strict": true as default.
"esModuleInterop" is always true and can't be set to false.

**Comment by rkrisztian** [comment]
Yes, @jramstedt, in the process of upgrading to TypeScript 6, I removed settings that are now defaults like `"strict": true`, and Next.js now does this to my `tsconfig.json`:

```none
  We detected TypeScript in your project and reconfigured your tsconfig.json file for you. Strict-mode is set to false by default.
  The following suggested values were added to your tsconfig.json. These values can be changed to fit your project's needs:

  	- strict was set to false

  Running TypeScript  ...Failed to type check.

Type error: Option 'exactOptionalPropertyTypes' cannot be specified without specifying option 'strictNullChecks'.
```

So now `"exactOptionalPropertyTypes": true` is broken too, unless I explicitly add back `"strict": true`.

I don't really like Next.js overriding my settings. Why not just leave this problem to new project generation or migration guides? Setting `"strict": false` shouldn't be an encouraged Next.js default. I have been using `"strict": true` for years, also in Next.js projects. At least I'd like to have a config flag that disallows Next.js from overriding my `tsconfig.json` settings.

**Comment by rkrisztian** [comment]
These are the settings that are now defaults but Next.js would force them to be set:

```none
    "strict": true,
    "module": "esnext",
    "esModuleInterop": true,
    "isolatedModules": true
```

More info on new defaults: https://devblogs.microsoft.com/typescript/announcing-typescript-6-0/#simple-default-changes

*Edit:* This migration guide could also be very helpful, for example: "`dom.iterable` and `dom.asynciterable` Merged into `dom`":
https://gist.github.com/privatenumber/3d2e80da28f84ee30b77d53e1693378f
